crates/sui-storage/src/fullnode_pending_tx_log.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! FullNodePendingTransactionLog is Write Ahead Log uesd in Transaction Orchestrator
+//! for transaction submission processing. It plays two roles:
+//! 1. At one time, a transaction is only processed once.
+//! 2. When Fullnode crashes and restarts, the pending transaction will be loaded and retried.
+
+use crate::mutex_table::MutexTable;
+use crate::write_ahead_log::{MUTEX_TABLE_SHARD_SIZE, MUTEX_TABLE_SIZE};
+use std::path::PathBuf;
+use sui_types::base_types::TransactionDigest;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::{TrustedTransction, VerifiedTransaction};
+use typed_store::traits::TypedStoreDebug;
+use typed_store::{rocks::DBMap, traits::Map};
+use typed_store_derive::DBMapUtils;
+
+pub type IsFirstRecord = bool;
+
+#[derive(DBMapUtils)]
+struct FullNodePendingTransactionTable {
+    logs: DBMap<TransactionDigest, TrustedTransction>,
+}
+
+pub struct FullNodePendingTransactionLog {
+    pending_transactions: FullNodePendingTransactionTable,
+    mutex_table: MutexTable<TransactionDigest>,
```

Comment:
this is a perfect use case for optimistic txns!

Reply:
Is the suggestion to keep the `MutexTable` here and let it be swapped with optimistic txn ?

> a plain old mutex might be good enough for this use case
Did we add `MutexTable` because `Mutex` hurts performance when the lock is held for long time?

Reply:
> Did we add MutexTable because Mutex hurts performance when the lock is held for long time?

More or less, yes. MutexTable is intended to allow for arbitrarily low lock contention when there are many cores that potentially hold locks for a long time (as we do when executing a tx). Using a plain mutex for transaction execution would effectively make the validator single-threaded.

In this case, unless you think that modifying the table itself will consume the bulk of the cpu time in this code, I think it is unlikely we're going to have much contention even on a single lock. But I think the typical request here will spend most of its time waiting on QD or local execution.

Still, MutexTable won't hurt you, its just probably overkill.